pkg/cluster/internal/providers/docker/network.go

```diff
@@ -0,0 +1,48 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package docker
+
+import (
+	"sigs.k8s.io/kind/pkg/exec"
+)
+
+// TODO: we'll probably allow configuring this
+//
+// however currently picking a single network is equivilant to the previous
+// behavior *except* that we moved from the default bridge to a user defined
+// network because the default bridge is actually special versus any other
+// docker network and lacks the emebdded DNS
+//
+// for now this also makes it easier for apps to join the same network
+const fixedNetworkName = "kind"
+
+// ensureNetwork checks if docker network by name exists, if not it creates it
+func ensureNetwork(name string) error {
+	out, err := exec.Output(exec.Command(
+		"docker", "network", "ls",
+		"--filter=name=^"+name,
+		"--format={{.Name}}",
+	))
+	if err != nil {
+		return err
+	}
+	// network already exists
+	if string(out) == name+"\n" {
+		return nil
+	}
+	return exec.Command("docker", "network", "create", "-d=bridge", name).Run()
```

Comment:
TODO

```go
if ipv6 {
exec.Command("docker", "network", "create", "--ipv6","-d=bridge", name).Run()
```

Reply:
can we add a comment?

Reply:
I'll probably actually fix it :sweat_smile:

Reply:
ohhhh ugh, if we want to support simultaneous ipv4 and ipv6 we're going to have to either:
- always enable ipv6 on the kind bridge
- require that users ensure the bridge exists w/ ipv6 before kind
- require that users create an ipv6 cluster first before any ipv4
- use a special network for ipv6 clusters

requiring that the ipv6 one is created first seems sanest?
that or having a `kind-ipv6` network ...

always enabling ipv6 won't work because docker OOTB will cause:
> Error response from daemon: could not find an available, non-overlapping IPv6 address pool among the defaults to assign to the network

due to ipv6 not being configured on the daemon.

Reply:
i implemented:
> require that users create an ipv6 cluster first before any ipv4

just to see how far CI gets.

Reply:
can we just retry and fallback to an ipv4 only bridge?

Reply:
not sure what you mean.
ipv6 has to be enabled at network creation time.

the old default one has it IF the daemon has ipv6 configured. but no embedded DNS.

if the user creates a kind cluster with ipv4 then creates ipv6 we'll have created a v4 only bridge.

we could try to check if the daemon has v6 configured when creating v4 clusters and opportunistically enable v6 maybe.

let's see how CI does first.

Reply:
```go

if err:=  exec.Command("docker", "network", "create", "--ipv6","-d=bridge", name).Run(): err != nil {
 return exec.Command("docker", "network", "create", "-d=bridge", name).Run()
}
return nil
```

Reply:
this will report odd errors when the user was actually trying to do ipv6.
I will think on this some more and follow up